Improve logging

williamlardier opened the pull request:
- Change the "received request" log level
- Better define the default logging fields
- Remove useless logs
- Log in case of error in the mongoclientinterface, as InternalError hides all the underlying information
- Pass the right error in case of error during deletion, to actually fix a bug in case of mongodb erroor leading to no error log AND deletion of data from the storage layer, but not in mongodb
- In the dataWrapper some logs were in trace, meaning no info available in case of error when handling data, and harder debugging
- Adding multiple tests missing today, affecting code coverage due to the logging updates
- Updated the in memory data store, as we need the `head` method to properly unit test backbeat routes.

Issue: ARSN-471

Seragonia committed:
Improve S3 logging

- No need to log incoming requests with info level as long as
  we log the response.
- Removed useless logs.
- Added missing logs.
- Support better logging in case of error in delete object
  (mongodb backend).
- Rework default fields.

Issue: ARSN-471

bert-e commented:
# Hello williamlardier,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/arsenal/doc/user).


<details>
  <summary><b>Available options</b></summary>

  name   | description  | privileged   | authored
  ------ | ------------ | ------------ |----------
  `/after_pull_request` | Wait for the given pull request id to be merged before continuing with the current one. |  | 
  `/bypass_author_approval` | Bypass the pull request author's approval |  :star:  | 
  `/bypass_build_status` | Bypass the build and test status |  :star:  | 
  `/bypass_commit_size` | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 
  `/bypass_incompatible_branch` | Bypass the check on the source branch prefix |  :star:  | 
  `/bypass_jira_check` | Bypass the Jira issue check |  :star:  | 
  `/bypass_peer_approval` | Bypass the pull request peers' approval |  :star:  | 
  `/bypass_leader_approval` | Bypass the pull request leaders' approval |  :star:  | 
  `/approve` | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
  `/create_pull_requests` | Allow the creation of integration pull requests. |  | 
  `/create_integration_branches` | Allow the creation of integration branches. |  | 
  `/no_octopus` | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
  `/unanimity` | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
  `/wait` | Instruct Bert-E not to run until further notice. |  | 
  

</details>



<details>
  <summary><b>Available commands</b></summary>

  name   | description  | privileged
  ------ | ------------ | ------------
  `/help` | Print Bert-E's manual in the pull request. | 
  `/status` | Print Bert-E's current status in the pull request ```TBA``` | 
  `/clear` | Remove all comments from Bert-E from the history ```TBA``` | 
  `/retry` | Re-start a fresh build ```TBA``` | 
  `/build` | Re-start a fresh build ```TBA``` | 
  `/force_reset` | Delete integration branches & pull requests, and restart merge process from the beginning. | 
  `/reset` | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
  
</details>




*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ARSN-471 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.1.148`

* `8.2.4`


Please check the `Fix Version/s` of ARSN-471, or the target
branch of this pull request.

Seragonia committed:
Improve auth logs

Issue: ARSN-471

Seragonia committed:
Improve dataWrapper log levels

Issue: ARSN-471

Seragonia committed:
Add tests and improve asserts

Issue: ARSN-471 and ARSN-59

Seragonia committed:
Add head method to inmem datastore

This is needed to unit test the backbeat routes

Issue: ARSN-471

welansari commented on tests/unit/storage/metadata/mongoclient/MongoClientInterface.spec.js:
```diff
@@ -846,6 +846,80 @@ describe('MongoClientInterface, tests', () => {
+            });
+        });
+
+        afterEach(async () => {
+            await new Promise(resolve =>
+                client.deleteBucket(bucketName, logger, err => {
+                    if (err) {
+                        logger.error('Failed to delete bucket in cleanup', { error: err });
+                    }
+                    resolve();
+                }),
+            );
+        });
```
should use callbacks like we do in the beforeEach or make both async ?

williamlardier replied:
I kept cb indeed.

Note: I also added a commit for the error handling in mongoclientinterface, it was in the same commit as logging, while it sis a logic change...

welansari commented on tests/unit/s3middleware/prepareStream.spec.js:
```diff
@@ -0,0 +1,31 @@
+        stream.headers['x-amz-content-sha256'] = 'STREAMING-AWS4-HMAC-SHA256-PAYLOAD';
+    });
+
+    it('should return the stream is sha256 header is not streaming', () => {
```
```suggestion
    it('should return the stream if sha256 header is not streaming', () => {
```

Seragonia committed:
Fix missing error handling in internalDeleteObject

- In case of error, no error was returned.
- As a result, we would continue the API and don't
  detect anything, which could lead to a deletion
  of storage data, but not in MongoDB.

Issue: ARSN-471

francoisferrand commented on lib/auth/Vault.ts:
```diff
@@ -1,4 +1,4 @@
-import { Logger } from 'werelogs';
+import { Logger, RequestLogger } from 'werelogs';
```
`Logger` not needed anymore

francoisferrand commented on lib/auth/Vault.ts:
```diff
@@ -445,7 +440,7 @@ export default class Vault {
     getOrCreateEncryptionKeyId(
         canonicalID: string,
-        log: Logger,
+        log: RequestLogger,
```
should update the comment as well (line 429: `@param {Logger} log - logger`)

francoisferrand commented on lib/auth/v4/streamingV4/V4Transform.ts:
```diff
@@ -1,7 +1,7 @@
 import async from 'async';
 import errors from '../../../errors';
-import { Logger } from 'werelogs';
+import { Logger, RequestLogger } from 'werelogs';
```
`Logger` not used anymore

francoisferrand commented on lib/s3routes/routes.ts:
```diff
@@ -221,17 +235,18 @@ export default function routes(
     }
 
-    log.end().addDefaultFields(clientInfo);
+    log.addDefaultFields(clientInfo);
```
removing `end()` means we may not see any log until the request is complete: which may make it harder to troubleshoot requests which somehow do not terminate. Is this really what we want?

williamlardier replied:
The `end` function in werelogs should only be used to end the RequestLogger. It returns an `EndLogger` that should only be used to log one last time. It also allows to have the "elapsed time" info. But as long as we log using the current logging level, the logs will be flushed in real time. When the dump level is reached, we might flush those logs again, but with more info (like, debug logs).

I am not sure to understand what you mean by "we may not see any log until the request is complete"?

francoisferrand replied:
> as long as we log using the current logging level, the logs will be flushed in real time

this answers the "we may not see any log until the request is complete" issue, so no problem actually :-)

francoisferrand commented on lib/s3routes/routesUtils.ts:
```diff
@@ -195,12 +193,10 @@ export const JSONResponseBackend = {
         response.writeHead(200, { 'Content-type': 'application/json' });
-        log.debug('response http code', { httpCode: 200 });
-        log.trace('json response', { json });
+        log.trace('sending success json response', { json });
```
* should we not keep the `httpCode` field? (may be consistent with other error logs)
* my understanding of the "duplicated" traces is that we want(ed) a debug log in anycase, but include JSON only when log level is trace : should we try to keep such behavior (i.e. changing the behavior based on `logger.config.level()`), or do we really _want_ to remove the debug log?

williamlardier replied:
We have the debug log when the response is complete just below, with the real status code (while the raw 200 is incorrect some times, could be 2xx). Meaning that, in case of debug log level, we end up with both the 200 and the real code sent to the client, which is not useful. This is removed on purpose, this brings 0 benefit as a log.

Then we keep a trace as before with the json info.

If we have an error log we should not enter the okResponse anyway (or it's a multi delete).

francoisferrand commented on tests/unit/storage/metadata/mongoclient/MongoClientInterface.spec.js:
```diff
@@ -846,6 +846,78 @@ describe('MongoClientInterface, tests', () => {
+
+            // Mock findOne to throw an error
+            const originalFindOne = collection.findOne;
+            collection.findOne = () => Promise.reject(new Error('Simulated MongoDB error'));
```
working on mongodb driver update, @benzekrimaha had issues when Promises where used to catch errors in mongodb driver, which AFAIU could only be solved by actually using try/catch block instead... and which could not be seen with such mocking of mongodb, but only with real mongodb integration tests.

not really related to this PR, but may be best to confirm ASAP that we are safe on that aspect (i.e. make sure we handle errors properly, and that we are not adding false confidence by adding tests where the mock does not behave like the real driver)

benzekrimaha replied:
When working on the bump of the driver on vault I indeed was able to see that without a try a catch that awaits for the operation , it was not working as expected as the operations were failing without being fulfilled.

benzekrimaha replied:
it was failing with no error thrown as it was not waiting for the operation , I only had the operations really failing / being processed when using a try/catch and an await on the operations

williamlardier replied:
Ok, I will try to review if we are missing something here, as this is not consistent with MongoDB documentation

williamlardier replied:
After reviewing Vault code, the issue is more on the lack of synchronous error handling: when it happens, we don't call mongodb indeed, because we are not ensured that the db is available before we call mongodb methods. However, this is only in Vault. Independently to how we implement, the mongodb driver will always call its promises properly.

Here, the tests are mocking a mongodb error, which is indeed not a synchronous error, but still tests that we properly handle mongodb errors. This is also confirmed by all CIs: if it was not working, nothing would pass.

The problem in vault is more a migration issue, where we switch from all-callbacks to promises, but the architecture is different and we are likely missing `catch` for some types of errors, or misconfigured how we use the promises.

I am closing this comment: in Arsenal, the migration was already done, and everything works as expected.

francoisferrand commented:
as a side note / complementary effort : we are changing some log levels here, but we should remember the logs (esp. debug/trace) are used in 2 contexts:

* Either when the user activate more verbose logging
  → display all the time, and indeed we may need to streamline these
* In case of error, every log related to the request is dumped again
  → in that specific context, having more (and separate lines) logs may/could be acceptable;
  → yet there may be an additional improvement in werelogs, to ensure these logs (some of which may already have been logged!) are clearly isolated: maybe by including them in the "final" json line or adding some sort of prefix/..., instead of just printing them like regular logs

That is probably a separate improvement, but sharing as I think it may bring a different point of view on the updating of logs.

Seragonia committed:
Remove unused import types

Issue: ARSN-471

ghost commented:
> → yet there may be an additional improvement in werelogs, to ensure these logs (some of which may already have been logged!) are clearly isolated: maybe by including them in the "final" json line or adding some sort of prefix/..., instead of just printing them like regular logs

Indeed today we might log twice the same logs, in particular those with logging level >= to the log level, if a dump occurs.

Note: there is also a performance impact of having logs under the levels we listen to, because we are still processing them "in case of error", and that might lead to a lot of object copy/parsing, even if not useful.

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ARSN-471 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.1.148`

* `8.2.5`


Please check the `Fix Version/s` of ARSN-471, or the target
branch of this pull request.

ghost commented:
Removing the last commit as we need to merge #2286  before releasing.

williamlardier commented:
/approve

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ARSN-471 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.1.148`

* `8.2.5`


Please check the `Fix Version/s` of ARSN-471, or the target
branch of this pull request.




*The following options are set:* **approve**

ghost commented:
/create_integration_branches

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/ARSN-471` into
`development/8.1`
* [w/8.2/improvement/ARSN-471](https://github.com/scality/arsenal/compare/development/8.2...w/8.2/improvement/ARSN-471) will be merged into `development/8.2` 



The following branches will **NOT** be impacted:

* `development/6.4`
* `development/7.10`
* `development/7.4`
* `development/7.70`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **create_integration_branches**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers











*The following options are set:* **create_integration_branches**

ghost commented:
/approve

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/8.1`

* :heavy_check_mark: `development/8.2`



The following branches have **NOT** changed:

* `development/6.4`
* `development/7.10`
* `development/7.4`
* `development/7.70`



Please check the status of the associated issue ARSN-471.

Goodbye williamlardier.




*The following options are set:* **approve, create_integration_branches**
